selene-lib/src/rules/mismatched_arg_count.rs

```diff
@@ -0,0 +1,449 @@
+            },
+            // function(a, b, ...) - if we call it through foo(a), b and the varargs could be implicitly nil.
+            // there is no upper bound since foo(a, b, c, d) is valid - therefore any amount of arguments provided is valid
+            ParameterCount::Minimum(_) => true,
+            // Any amount of arguments could be provided
+            ParameterCount::Variable => true,
+        }
+    }
+
+    pub fn to_message(self, provided: PassedArgumentCount) -> String {
+        match self {
+            ParameterCount::Fixed(required) => {
+                format!(
+                    "this function takes {} {} but {} were supplied",
+                    required,
+                    plural(required, "argument", "arguments"),
+                    provided
+                )
+            }
+            ParameterCount::Minimum(required) => format!(
+                "this function takes at least {} {} but {} were supplied",
+                required,
+                plural(required, "argument", "arguments"),
+                provided
+            ),
+            ParameterCount::Variable => "a variable amount of arguments".to_owned(),
+        }
+    }
+}
+
+impl Display for ParameterCount {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            ParameterCount::Fixed(required) => write!(
+                f,
+                "expected {} {}",
+                required,
+                plural(*required, "argument", "arguments")
+            ),
+            ParameterCount::Minimum(required) => {
+                write!(
+                    f,
+                    "expected at least {} {}",
+                    required,
+                    plural(*required, "argument", "arguments")
+                )
+            }
+            ParameterCount::Variable => write!(f, "expected any number of arguments"),
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+enum PassedArgumentCount {
+    /// Passed a fixed amount of arguments, such as foo(a, b, c) or foo(a, call(), c) or foo(a, ..., c)
+    Fixed(usize),
+    /// Passed a variable of arguments - but we know the lower bound: e.g. foo(a, b, call()) or foo(a, b, ...)
+    Variable(usize),
+}
+
+fn is_vararg(expression: &ast::Expression) -> bool {
+    if let ast::Expression::Value { value, .. } = expression {
+        if let ast::Value::Symbol(token) = &**value {
+            if let tokenizer::TokenType::Symbol { symbol } = token.token().token_type() {
+                if let tokenizer::Symbol::Ellipse = symbol {
+                    return true;
+                }
+            }
+        }
+    }
+
+    false
+}
+
+impl PassedArgumentCount {
+    fn from_function_args(function_args: &ast::FunctionArgs) -> Self {
+        match function_args {
+            ast::FunctionArgs::Parentheses { arguments, .. } => {
+                // We need to be wary of items with side effects, such as function calls or ... being the last argument passed
```

Comment:
Thinking it over, `has_side_effects` is not what you want here. You just want to check if something is a function call. `{ foo() }`, for instance, has `has_side_effects() == true`, but it doesn't make any difference as to whether something changes variable arguments..